### PR TITLE
fixes #3751 - new/edit submit buttons now have id

### PR DESCRIPTION
**Pull request by lzap**
This adds HTML id attributes to all submit buttons created with
submit_or_cancel function. QA engineers requested that.

@omaciel can you test this patch if it is sufficient for you?

It should cover 95 % of all cases. If this works for you and you find some
remaining forms where you want to add id attributes, open another issue and
list all the elements there, I will add the rest manually.

I was unsure where to put a unit/functional test, so I used architectures for
that which works fine.

### CODE REVIEW

**Comment by domcleal** [comment]
More comments on http://projects.theforeman.org/issues/3751

**Comment by lzap** [comment]
So I am rebasing with additional change. In addition to input submit button, we now put id attributes to _all_ hyperlinks generated with link_to method if and only if id is missing. That is almost all of them. I generate them from HREF attribute - last three elements of the path are converted with a prefix. Examples:

```
<a href="/architectures/new" id="id_architectures_new" class="btn btn-success">New Architecture</a>

<a href="/architectures" class="btn btn-default two-pane-close" id="aid_architectures">Cancel</a>

<a href="/hosts/test3.local.lan" id="aid_hosts_test3.local.lan"><span>test3.local.lan</span></a>

<a href="/hosts/test1.local.lan/edit" id="aid_hosts_test1.local.lan_edit">Edit</a>
```

Form submit input buttons have different convention, I am generating this from form name directly (no need to parse HREF in this case):

```
<input class="btn btn-primary" id="aid_create_subnet" name="commit" type="submit" value="Submit">

<input class="btn btn-primary" id="aid_update_domain" name="commit" type="submit" value="Submit">
```

I am overriding link_to helper method and adding id prefixed with "aid" (autogenerated id). Results look sane from what I have seen.

Og please test this carefuly, once you commit to a id generating solution, we do not want to change that (as you'd need to rewrite all your automation again).

**Comment by omaciel** [comment]
@lzap will try it this morning

**Comment by omaciel** [comment]
Tested and loved it!!!  +1

**Comment by lzap** [comment]
Okay, @omaciel please give me more feedback, try to automate few things etc until our review is done and I have chance to incorporate your possible remarks. I am going to fix unit tests and this is open for review!

**Comment by lzap** [comment]
Tests are done, @omaciel waiting for your final go! :-)

**Comment by domcleal** [comment]
@abenari could you review this please?

It doesn't worry me too much - I like that you've added rescue blocks in, since it's a core piece of code.

**Comment by abenari** [comment]
In some pages we have multiple links with the same href. This is useful practice in UI design because different people look for the actions in different places. This patch will create duplicated ID in this case (http://programmers.stackexchange.com/questions/127178/two-html-elements-with-same-id-attribute-how-bad-is-it-really).

The patch is ok as is, if @omaciel can work with html5 data selectors I would prefer to have 'data-id' attribute instead of id in the links, but this isn't critical.

**Comment by lzap** [comment]
> In some pages we have multiple links with the same href. This is
> useful practice in UI design because different people look for the
> actions in different places. This patch will create duplicated ID in
> this case

Yes, that was the reason why I created the "aid" prefix (autogenerated
id) so if anyone is bothered with this, he can just add an id to any of
these links and id will not be generated from that point.

> The patch is ok as is, if @omaciel can work with html5 data selectors
> I would prefer to have 'data-id' attribute instead of id in the links,
> but this isn't critical.

So you would like to have the id in "data-id" instead of "id"? Easy
change for me I guess. @omaciel?

Thanks Amos for your review.

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

**Comment by omaciel** [comment]
@lzap sure, `data-id` also works, as long as they're unique :)

**Comment by lzap** [comment]
Ok I refactored this to use "data-id". Ogg please do final testing and once you waive it - good to merge.

**Comment by blrm** [comment]
@omaciel asked me to test this, and the data-id attributes look good. These will work for our use cases. Thanks!

**Comment by lzap** [comment]
Great, thanks! Looking good.

File: app/helpers/layout_helper.rb [hunk 1]
```diff
@@ -158,11 +158,23 @@ def help_inline(inline, error)
     end
   end
 
+  def form_to_submit_id f
+    object = f.object.respond_to?(:to_model) ? f.object.to_model : f.object
+    key = object ? (object.persisted? ? :update : :create) : :submit
+    model = if object.class.respond_to?(:model_name)
+              object.class.model_name.human.downcase
```
Comment (domcleal) [inline app/helpers/layout_helper.rb:167]: This has some odd characteristics, and I'm unsure how much you intended to munge the IDs generated.  e.g. on the new host form, you get `aid_create_host/managed` (with a slash) and on compute resources you get `aid_update_compute resource` (with a space).

If you have mark translations enabled too, you get `aid_update_xcompute resourcex`.  I'm unsure if QE will use this feature?

**Comment by omaciel** [comment]
The generated IDs _must_ be unique and not change with localization, otherwise we won't be able to perform our functional tests with different locales.

File: app/helpers/layout_helper.rb (hunk 1 above)
Comment (lzap) [inline app/helpers/layout_helper.rb:167]: Ok that's right, I did not test with hosts and spaces.

Both fixed, translations does not change the generated IDs.

**Comment by domcleal** [comment]
Merged as bfee97b for Foreman 1.5.0, thanks @lzap!

**Pull request closed without merging**